File: python/stack/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        st = []
        data = []

        if root:
            st.append(root)

        while st:
            cur = st.pop()
            data.append(cur.val)
            if cur.right:
                st.append(cur.right)
            if cur.left:
                st.append(cur.left)

        return data

File: c/stack/binary-tree-preorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    struct TreeNode **st;
    struct TreeNode *cur;
    int *data;
    int idx = 0;
    int top = -1;
    int data_count = 32;
    int st_count = 32;

    st = malloc(sizeof(struct TreeNode *) * st_count);
    data = malloc(sizeof(int) * data_count);

    if(root)
        st[++top] = cur;

    while(top > -1){
        cur = st[top--];

        data[idx++] = cur->val;
        if(idx == data_count){
            data_count <<= 1;
            data = realloc(data, sizeof(int) * data_count);
        }

        if(cur->right)
            st[++top] = cur->right;

        if(cur->left)
            st[++top] = cur->left;

        if(top+1 == st_count){
            st_count <<= 1;
            st = realloc(st, sizeof(struct TreeNode *) * st_count);
        }
    }

    *returnSize = idx;

    free(st);
    return data;
}
